orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,208 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
   graph.AddInitializedTensor(proto_data);
 }
 
+Status FindAllConnectedNodes(Graph& graph,
+                             const Node* node,
+                             std::vector<const Node*>& connected_nodes,
+                             std::set<const NodeArg*>& connected_inputs,
+                             std::set<const NodeArg*>& connected_outputs
+                             ) {
+  ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+      node->InputDefs(),
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (graph.IsInputsIncludingInitializers(&node_arg)) {
+          connected_inputs.insert(&node_arg);
+        } else {
+          const Node* producer_node = graph.GetProducerNode(node_arg.Name());
+          if (producer_node == nullptr) {
+            // got nullptr as producer node. This could be because the input is a constant op which will be optimized
+            // away. Print out this information and continue.
+            // LOGS_DEFAULT(WARNING) << "Cannot find producer node for node_arg: " << node_arg.Name() << ". Skipping this node.";
+          } else {
+            connected_nodes.push_back(producer_node);
+          }
+        }
+        return Status::OK();
+      }));
+
+  ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+      node->OutputDefs(),
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (!graph.IsOutput(&node_arg)) {
+          std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(node_arg.Name());
+          connected_nodes.insert(std::end(connected_nodes), consumer_nodes.begin(), consumer_nodes.end());
+
+        } else {
+          connected_outputs.insert(&node_arg);
+        }
+        return Status::OK();
+      }));
+  return Status::OK();
+}
+
+// PipelineStageNodeGroup groups nodes that share the same input initializer and belong to the same stage.
+// It is used to distinguish other nodes that share the same input initializer but belong to
+// other pipeline partitions after split.
+struct PipelineStageNodeGroup {
+  const size_t stage_id;
+
+  // Vector of nodes that have the same initializer input and belong to the same stage. Noted that
+  // the consumer nodes of a particular initializer can be more than one, so we need a vector to store those
+  // nodes.
+  std::vector<Node*> nodes;
+  PipelineStageNodeGroup(const size_t stage, std::vector<Node*>& node) : stage_id(stage), nodes(std::move(node)){};
+};
+
+common::Status AddPassthroughInitializer(Graph& graph,
+                                         const NodeArg* initializer,
+                                         const std::vector<PipelineStageNodeGroup>& node_groups,
+                                         const std::vector<Node*>& send_nodes,
+                                         const std::vector<Node*>& recv_nodes) {
+  ORT_ENFORCE(node_groups.size() >= 2, "Initializer ", initializer->Name(),
+              " is not shared across stages. It only exits in partition: ", node_groups[0].stage_id);
```

Comment:
Is it possible that `node_groups` could be empty here (and so `node_groups[0]` could be out of bounds)? (I'm not sure if this error check is more of an assertion/internal integrity check, or if the user can specify invalid input that would trigger this check.)

Reply:
This method is called after a check to ensure node_groups's size is greater than 1 (see line 1013 `if (entry.second.size() > 1)` in `HandleSharedInitializer`. This check is an internal check for future development to not mis-use this method when new function calls this method -- in other word, user won't be able to specify input that triggers this check :-)

But I'll change to avoid the use of node_groups[0], as this creates confusion by looking at this check. Thanks for pointing out. 

---
In reply to: [449303044](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r449303044) [](ancestors = 449303044)